clustering/singleton/server/src/main/java/org/wildfly/clustering/singleton/server/DistributedSingletonServiceBuilder.java

```diff
@@ -59,17 +62,14 @@ public void accept(V value) {
 
     @Override
     public SingletonServiceBuilder<T> setInstance(Service service) {
-        this.context.setServiceName(ServiceName.parse(service.getClass().getName()));
         this.context.setService(service);
         return this;
     }
 
     @Override
     public SingletonServiceController<T> install() {
-        ServiceName name = this.context.getServiceName();
-        if (name == null) {
-            throw new IllegalStateException();
-        }
+        // Identify singleton service via its id if caller did not provide one
+        this.context.setServiceName(this.id);
```

Comment:
Note to reviewers/maintainers: The 'if caller did not provide one' bit is in the setServiceName implementation.